documentation/design-docs/diagnostics-client-library.md

```diff
@@ -0,0 +1,329 @@
+    };
+
+    EventPipeSession counterSession = DiagnosticsClient.StartTracing(processId, runtimeCounterProvider);
+    EventPipeEventSource source = new EventPipeEventSource(counterSession.stream)
+    source.Dynamic.All += (TraceEvent obj) => {
+        if (obj.EventName.Equals("EventCounters"))
+        {
+            // I know this part is ugly. But this is all TraceEvent.
+            IDictionary<string, object> payloadFields = (IDictionary<string, object>)(obj.GetPayloadValueByName("Payload"));
+            if (payloadFields["Name"].ToString().Equals("cpu-usage"))
+            {
+                double cpuUsage = Double.Parse(payloadFields["Mean"]);
+                if (cpuUsage > (double)threshold)
+                {
+                    EventPipeSession traceSession = DiagnosticsClient.StartTracing(processId, cpuProvider);
+                    if (traceSesssion != null)
+                    {
+                        await Task.Run(() => {
+                            var buffer = new byte[16 * 1024];
+                            var fs = new FileStream(traceName, FileMode.Create, FileAccess.Write))
+                            Stopwatch sw = new Stopwatch();
+                            sw.Start();
+                            while (sw.Elapsed.Seconds < traceDuration)
+                            {
+                                int nBytesRead = traceSession.Stream.Read(buffer, 0, buffer.Length);
+                                if (nBytesRead <= 0)
+                                    break;
+                                fs.Write(buffer, 0, nBytesRead);
+                            }
+                            DiagnosticsClient.StopTracing(traceSession);
+                            shouldExit.Set();
+                        });
+                    }
+                }
+            }
+        }
+    }
+    try
+    {
+        source.Process();
+        shouldExit.WaitOne();
+    }
+    catch (Exception e) { } 
+    finally
+    {
+        DiagnosticsClient.StopTracing(counterSession);
+    }
+}
+```
+
+## API Descriptions
+
+At a high level, the DiagnosticsClient class provides static methods that the user may call to invoke diagnostics IPC commands (i.e. start an EventPipe session, request a core dump, etc.) The library also provides several classes that may be helpful for invoking these commands. These commands are described in more detail in the diagnostics IPC protocol documentation available here: https://github.com/dotnet/diagnostics/blob/master/documentation/design-docs/ipc-protocol.md#commands. 
+
+### ProcessLocator
+This is a helper class that finds processes to attach to.
+```cs
+namespace Microsoft.Diagnostics.NETCore.Client
+{
+    /// <summary>
+    /// A utility class that contain various helper methods to interact with the diagnostics IPC.
+    /// </summary>
+    public class ProcessLocator
+    {
```

Comment:
We currently only look in `/tmp` on Linux, do we want to expand this to allow for the user to specify a directory to look for sockets in?  I'm in thinking about the container scenarios where we mount the container's `/tmp` to the host's.  If we make the search directory arbitrary, we could mount `/tmp` to wherever and the tools would still work.

Reply:
We can add that in later unless it's in a dire need?